File: c/src/integration_geo.c
/*
 * integration_geo.c — CircleAI.Integration.Geo (C11 port).
 *
 * In-memory IWeatherProvider (open-meteo) + IRoutingProvider (osrm). The real
 * providers hit Open-Meteo / OSRM; here samples + route are synthesised
 * deterministically from the inputs. The load-bearing pure logic — the WMO
 * weather-code decode table and the OSRM mode->profile mapping — is ported
 * exactly. Pure C11 + libc + libm. No pthreads.
 */

#include "circle_ai/integration_geo.h"
#include "board_common.h"
#include <math.h>

/* ── ported pure logic ──────────────────────────────────────────────────── */

const char *ca_int_open_meteo_wmo_decode(int code) {
    switch (code) {
        case 0:  return "clear sky";
        case 1: case 2: case 3: return "partly cloudy";
        case 45: case 48: return "fog";
        case 51: case 53: case 55: return "drizzle";
        case 56: case 57: return "freezing drizzle";
        case 61: case 63: case 65: return "rain";
        case 66: case 67: return "freezing rain";
        case 71: case 73: case 75: return "snow";
        case 77: return "snow grains";
        case 80: case 81: case 82: return "rain showers";
        case 85: case 86: return "snow showers";
        case 95: return "thunderstorm";
        case 96: case 99: return "thunderstorm with hail";
        default: return "unknown";
    }
}

const char *ca_int_osrm_profile(const char *mode) {
    if (!mode) return "driving";
    if (cab_ord_eq(mode, "bike") || cab_ord_eq(mode, "bicycle")) return "bike";
    if (cab_ord_eq(mode, "foot") || cab_ord_eq(mode, "walk"))    return "foot";
    return "driving";
}

/* ── weather provider (open-meteo) ──────────────────────────────────────── */

/* Deterministic surrogate weather-code from position (stands in for the network
 * value; stable per (lat,lon,hour) so results are reproducible). */
static int synth_code(double lat, double lon, int hour) {
    long h = (long)(fabs(lat) * 7.0 + fabs(lon) * 3.0) + hour;
    static const int codes[] = {0, 1, 2, 3, 45, 51, 61, 71, 80, 95};
    long i = h % 10;
    if (i < 0) i += 10;
    return codes[i];
}

static void synth_sample(double lat, double lon, int hour,
                         ca_int_weather_sample_t *s) {
    memset(s, 0, sizeof(*s));
    int code = synth_code(lat, lon, hour);
    double base = 15.0 + sin((lat + lon + hour) * 0.1) * 8.0;
    s->at_utc_ms    = (int64_t)hour * 3600000LL;
    s->temp_c       = base;
    s->feels_like_c = base - 1.5;
    s->precip_mm    = (code >= 51) ? 1.0 : 0.0;
    /* m/s surrogate * 3.6 -> km/h, mirroring the C# WindKph conversion. */
    s->wind_kph     = (2.0 + (double)((hour % 5))) * 3.6;
    s->cloud_pct    = (code == 0) ? 0 : (code < 45 ? 40 : 90);
    /* condition set by caller via wmo decode (kept centralised). */
    s->condition    = NULL;
}

static const char *weather_provider_id(void *impl) {
    (void)impl;
    return "open-meteo";
}

static int weather_current(void *impl, double lat, double lon,
                           ca_int_weather_sample_t *out) {
    (void)impl;
    if (!out) return -1;
    synth_sample(lat, lon, 0, out);
    out->condition = cab_strdup_empty(ca_int_open_meteo_wmo_decode(
        synth_code(lat, lon, 0)));
    if (!out->condition) { ca_int_weather_sample_free(out); return -1; }
    return 0;
}

static ca_int_weather_sample_t *weather_hourly(void *impl, double lat, double lon,
                                               int hours, size_t *out_count) {
    (void)impl;
    if (!out_count) return NULL;
    if (hours <= 0 || hours > 168) { *out_count = (size_t)-1; return NULL; }
    ca_int_weather_sample_t *out =
        (ca_int_weather_sample_t *)calloc((size_t)hours, sizeof(*out));
    if (!out) { *out_count = (size_t)-1; return NULL; }
    for (int i = 0; i < hours; ++i) {
        synth_sample(lat, lon, i, &out[i]);
        out[i].condition = cab_strdup_empty(ca_int_open_meteo_wmo_decode(
            synth_code(lat, lon, i)));
        if (!out[i].condition) {
            ca_int_weather_sample_free_array(out, (size_t)i);
            *out_count = (size_t)-1;
            return NULL;
        }
    }
    *out_count = (size_t)hours;
    return out;
}

ca_int_weather_provider_t *ca_int_open_meteo_create(void) {
    ca_int_weather_provider_t *p =
        (ca_int_weather_provider_t *)calloc(1, sizeof(*p));
    if (!p) return NULL;
    p->impl        = NULL; /* stateless */
    p->provider_id = weather_provider_id;
    p->current     = weather_current;
    p->hourly      = weather_hourly;
    return p;
}

void ca_int_weather_provider_destroy(ca_int_weather_provider_t *p) {
    free(p);
}

/* ── routing provider (osrm) ────────────────────────────────────────────── */

typedef struct {
    char *host; /* owned; parity only */
} routing_impl_t;

static const char *routing_provider_id(void *impl) {
    (void)impl;
    return "osrm";
}

/* Great-circle distance in metres (haversine). */
static double haversine_m(double lat1, double lon1, double lat2, double lon2) {
    const double R = 6371000.0;
    const double d2r = 3.14159265358979323846 / 180.0;
    double dlat = (lat2 - lat1) * d2r;
    double dlon = (lon2 - lon1) * d2r;
    double a = sin(dlat / 2) * sin(dlat / 2) +
               cos(lat1 * d2r) * cos(lat2 * d2r) * sin(dlon / 2) * sin(dlon / 2);
    return R * 2.0 * atan2(sqrt(a), sqrt(1.0 - a));
}

static int routing_route(void *impl, double from_lat, double from_lon,
                         double to_lat, double to_lon, const char *mode,
                         ca_int_route_estimate_t *out) {
    (void)impl;
    if (!out) return -1;
    memset(out, 0, sizeof(*out));

    /* Profile speed (m/s): driving ~13.9, bike ~4.2, foot ~1.4. */
    const char *profile = ca_int_osrm_profile(mode);
    double speed_ms = 13.9;
    if (cab_ord_eq(profile, "bike")) speed_ms = 4.2;
    else if (cab_ord_eq(profile, "foot")) speed_ms = 1.4;

    double dist_m = haversine_m(from_lat, from_lon, to_lat, to_lon);
    out->distance_km = dist_m / 1000.0;                    /* metres -> km */
    out->duration_ms = (int64_t)((dist_m / speed_ms) * 1000.0); /* seconds -> ms */

    out->polyline = (ca_int_route_point_t *)malloc(2 * sizeof(ca_int_route_point_t));
    if (!out->polyline) return -1;
    out->polyline[0].lat = from_lat;
    out->polyline[0].lon = from_lon;
    out->polyline[1].lat = to_lat;
    out->polyline[1].lon = to_lon;
    out->polyline_count = 2;
    return 0;
}

ca_int_routing_provider_t *ca_int_osrm_create(const char *host) {
    routing_impl_t *m = (routing_impl_t *)calloc(1, sizeof(routing_impl_t));
    if (!m) return NULL;
    m->host = cab_strdup_empty(host ? host : "https://router.project-osrm.org");
    if (!m->host) { free(m); return NULL; }

    ca_int_routing_provider_t *p =
        (ca_int_routing_provider_t *)calloc(1, sizeof(*p));
    if (!p) { free(m->host); free(m); return NULL; }
    p->impl        = m;
    p->provider_id = routing_provider_id;
    p->route       = routing_route;
    return p;
}

void ca_int_routing_provider_destroy(ca_int_routing_provider_t *p) {
    if (!p) return;
    routing_impl_t *m = (routing_impl_t *)p->impl;
    if (m) { free(m->host); free(m); }
    free(p);
}

File: go/integration_geo.go
// integration_geo.go
//
// Ports CircleAI.Integration.Geo:
//   OpenMeteoWeatherProvider -> OpenMeteoWeatherProvider (IWeatherProvider)
//   OsrmOptions / OsrmRoutingProvider -> OsrmOptions / OsrmRoutingProvider (IRoutingProvider)
//
// Both providers speak a real public HTTP API; the live HttpClient is replaced by
// the injected CarrierHTTP seam per the porting rules, so they are deterministic
// and make no network calls. Wire details (URL construction with invariant-culture
// coordinates, query params, JSON field extraction, the m/s→km/h and metre→km
// conversions, WMO decode, and the mode→profile mapping) are reproduced faithfully.

package circleai

import (
	"context"
	"errors"
	"strconv"
	"strings"
	"time"
)

// ── Open-Meteo weather ──────────────────────────────────────────────────────

// OpenMeteoWeatherProvider is the Open-Meteo (no-API-key) weather provider over
// the injected CarrierHTTP. Ports OpenMeteoWeatherProvider.
type OpenMeteoWeatherProvider struct {
	http CarrierHTTP
}

// NewOpenMeteoWeatherProvider constructs the provider. http is required (the C#
// ctor throws on null http).
func NewOpenMeteoWeatherProvider(http CarrierHTTP) (*OpenMeteoWeatherProvider, error) {
	if http == nil {
		return nil, errors.New("http is required")
	}
	return &OpenMeteoWeatherProvider{http: http}, nil
}

// ProviderID is "open-meteo".
func (p *OpenMeteoWeatherProvider) ProviderID() string { return "open-meteo" }

// Current ports CurrentAsync: fetch the current-conditions block and map it.
func (p *OpenMeteoWeatherProvider) Current(_ context.Context, lat, lon float64) (WeatherSample, error) {
	url := "https://api.open-meteo.com/v1/forecast?latitude=" + fmtFloat(lat) + "&longitude=" + fmtFloat(lon) +
		"&current=temperature_2m,apparent_temperature,precipitation,wind_speed_10m,cloud_cover,weather_code"
	resp, err := p.http.Do(&CarrierHTTPRequest{Method: "GET", URL: url})
	if err != nil {
		return WeatherSample{}, err
	}
	if !carrierHTTPStatusOK(resp.StatusCode) {
		return WeatherSample{}, statusError("Open-Meteo current", resp.StatusCode)
	}
	root, err := parseJSONObject(resp.Body)
	if err != nil {
		return WeatherSample{}, err
	}
	cur, ok := tjObject(root, "current")
	if !ok {
		return WeatherSample{}, errors.New("Open-Meteo response missing 'current'")
	}
	ts, _ := tjString(cur, "time")
	temp, _ := tjFloat(cur, "temperature_2m")
	feel, _ := tjFloat(cur, "apparent_temperature")
	precip, _ := tjFloat(cur, "precipitation")
	wind, _ := tjFloat(cur, "wind_speed_10m")
	cloud, _ := tjInt(cur, "cloud_cover")
	code, _ := tjInt(cur, "weather_code")
	return WeatherSample{
		AtUtc:      openMeteoTime(ts),
		TempC:      temp,
		FeelsLikeC: feel,
		PrecipMm:   precip,
		WindKph:    wind * 3.6,
		CloudPct:   cloud,
		Condition:  wmoDecode(code),
	}, nil
}

// Hourly ports HourlyAsync: fetch the hourly block and map min(len,hours) samples.
func (p *OpenMeteoWeatherProvider) Hourly(_ context.Context, lat, lon float64, hours int) ([]WeatherSample, error) {
	if hours <= 0 || hours > 168 {
		return nil, errors.New("hours out of range")
	}
	url := "https://api.open-meteo.com/v1/forecast?latitude=" + fmtFloat(lat) + "&longitude=" + fmtFloat(lon) +
		"&hourly=temperature_2m,apparent_temperature,precipitation,wind_speed_10m,cloud_cover,weather_code" +
		"&forecast_hours=" + itoaSmall(hours)
	resp, err := p.http.Do(&CarrierHTTPRequest{Method: "GET", URL: url})
	if err != nil {
		return nil, err
	}
	if !carrierHTTPStatusOK(resp.StatusCode) {
		return nil, statusError("Open-Meteo hourly", resp.StatusCode)
	}
	root, err := parseJSONObject(resp.Body)
	if err != nil {
		return nil, err
	}
	h, ok := tjObject(root, "hourly")
	if !ok {
		return nil, errors.New("Open-Meteo response missing 'hourly'")
	}
	timeArr, _ := tjArray(h, "time")
	temp, _ := tjArray(h, "temperature_2m")
	feel, _ := tjArray(h, "apparent_temperature")
	prec, _ := tjArray(h, "precipitation")
	wind, _ := tjArray(h, "wind_speed_10m")
	cld, _ := tjArray(h, "cloud_cover")
	code, _ := tjArray(h, "weather_code")

	n := minInt(len(timeArr), hours)
	result := make([]WeatherSample, 0, n)
	for i := 0; i < n; i++ {
		result = append(result, WeatherSample{
			AtUtc:      openMeteoTime(elemString(timeArr, i)),
			TempC:      elemFloat(temp, i),
			FeelsLikeC: elemFloat(feel, i),
			PrecipMm:   elemFloat(prec, i),
			WindKph:    elemFloat(wind, i) * 3.6,
			CloudPct:   elemInt(cld, i),
			Condition:  wmoDecode(elemInt(code, i)),
		})
	}
	return result, nil
}

// openMeteoTime parses an Open-Meteo timestamp (assume-UTC), falling back to
// nowUTC when absent — mirroring the C# `ts ?? DateTime.UtcNow.ToString("O")`
// then Parse(AssumeUniversal).ToUniversalTime().
func openMeteoTime(ts string) time.Time {
	if strings.TrimSpace(ts) == "" {
		return nowUTCFunc()
	}
	if t := parseDateTimeOffsetUTC(ts); !t.IsZero() {
		return t
	}
	return nowUTCFunc()
}

// wmoDecode decodes a WMO weather code to a condition string. Ports WmoDecode.
func wmoDecode(code int) string {
	switch code {
	case 0:
		return "clear sky"
	case 1, 2, 3:
		return "partly cloudy"
	case 45, 48:
		return "fog"
	case 51, 53, 55:
		return "drizzle"
	case 56, 57:
		return "freezing drizzle"
	case 61, 63, 65:
		return "rain"
	case 66, 67:
		return "freezing rain"
	case 71, 73, 75:
		return "snow"
	case 77:
		return "snow grains"
	case 80, 81, 82:
		return "rain showers"
	case 85, 86:
		return "snow showers"
	case 95:
		return "thunderstorm"
	case 96, 99:
		return "thunderstorm with hail"
	default:
		return "unknown"
	}
}

// ── OSRM routing ────────────────────────────────────────────────────────────

// osrmDefaultHost is the C# default OSRM host.
const osrmDefaultHost = "https://router.project-osrm.org"

// OsrmOptions configures the OSRM routing provider. Ports OsrmOptions. An empty
// Host defaults to the public demo server.
type OsrmOptions struct {
	Host string
}

// OsrmRoutingProvider is an OSRM HTTP client over the injected CarrierHTTP. Ports
// OsrmRoutingProvider.
type OsrmRoutingProvider struct {
	http CarrierHTTP
	opts OsrmOptions
}

// NewOsrmRoutingProvider constructs the provider. http is required; an empty Host
// defaults to osrmDefaultHost.
func NewOsrmRoutingProvider(http CarrierHTTP, opts OsrmOptions) (*OsrmRoutingProvider, error) {
	if http == nil {
		return nil, errors.New("http is required")
	}
	if opts.Host == "" {
		opts.Host = osrmDefaultHost
	}
	return &OsrmRoutingProvider{http: http, opts: opts}, nil
}

// ProviderID is "osrm".
func (p *OsrmRoutingProvider) ProviderID() string { return "osrm" }

// Route ports RouteAsync: map mode→profile, build the coordinate URL, require
// code=="Ok", and map routes[0] distance/duration/geometry.
func (p *OsrmRoutingProvider) Route(_ context.Context, fromLat, fromLon, toLat, toLon float64, mode string) (RouteEstimate, error) {
	profile := osrmProfile(mode)
	url := strings.TrimRight(p.opts.Host, "/") + "/route/v1/" + profile + "/" +
		fmtFloat(fromLon) + "," + fmtFloat(fromLat) + ";" +
		fmtFloat(toLon) + "," + fmtFloat(toLat) +
		"?overview=full&geometries=geojson"
	resp, err := p.http.Do(&CarrierHTTPRequest{Method: "GET", URL: url})
	if err != nil {
		return RouteEstimate{}, err
	}
	if !carrierHTTPStatusOK(resp.StatusCode) {
		return RouteEstimate{}, statusError("OSRM route", resp.StatusCode)
	}
	root, err := parseJSONObject(resp.Body)
	if err != nil {
		return RouteEstimate{}, err
	}
	code, _ := tjString(root, "code")
	if code != "Ok" {
		return RouteEstimate{}, errors.New("OSRM returned code=" + code)
	}
	routes, ok := tjArray(root, "routes")
	if !ok || len(routes) == 0 {
		return RouteEstimate{}, errors.New("OSRM returned no routes")
	}
	route, ok := asJSONObject(routes[0])
	if !ok {
		return RouteEstimate{}, errors.New("OSRM route[0] not an object")
	}
	dist, _ := tjFloat(route, "distance") // metres
	dur, _ := tjFloat(route, "duration")  // seconds
	poly := []GeoPoint{}
	if geom, ok := tjObject(route, "geometry"); ok {
		if coords, ok := tjArray(geom, "coordinates"); ok {
			for _, pt := range coords {
				arr, ok := pt.([]interface{})
				if !ok || len(arr) < 2 {
					continue
				}
				poly = append(poly, GeoPoint{Lat: toFloat(arr[1]), Lon: toFloat(arr[0])})
			}
		}
	}
	return RouteEstimate{
		DistanceKm: dist / 1000.0,
		Duration:   time.Duration(dur * float64(time.Second)),
		Polyline:   poly,
	}, nil
}

// osrmProfile maps a travel mode to an OSRM profile. Ports the switch.
func osrmProfile(mode string) string {
	switch mode {
	case "bike", "bicycle":
		return "bike"
	case "foot", "walk":
		return "foot"
	default:
		return "driving"
	}
}

// ── shared numeric/JSON helpers for geo ─────────────────────────────────────

// fmtFloat renders a float64 with the shortest round-trip representation using a
// '.' decimal point (InvariantCulture), matching double.ToString(InvariantCulture)
// for the finite values these providers emit.
func fmtFloat(v float64) string { return strconv.FormatFloat(v, 'g', -1, 64) }

// toFloat converts a decoded JSON value to float64 (0 on failure).
func toFloat(v interface{}) float64 {
	f, _ := tjFloatRaw(v)
	return f
}

// elemFloat/elemInt/elemString index a decoded JSON array with a scalar getter,
// returning the zero value when out of range or the wrong kind. These mirror the
// C# array-indexer + Get* calls over the parallel Open-Meteo arrays.
func elemFloat(arr []interface{}, i int) float64 {
	if i < 0 || i >= len(arr) {
		return 0
	}
	f, _ := tjFloatRaw(arr[i])
	return f
}

func elemInt(arr []interface{}, i int) int {
	if i < 0 || i >= len(arr) {
		return 0
	}
	n, _ := tjIntRaw(arr[i])
	return n
}

func elemString(arr []interface{}, i int) string {
	if i < 0 || i >= len(arr) {
		return ""
	}
	if s, ok := arr[i].(string); ok {
		return s
	}
	return ""
}

var (
	_ IWeatherProvider = (*OpenMeteoWeatherProvider)(nil)
	_ IRoutingProvider = (*OsrmRoutingProvider)(nil)
)
